quinn-h3/src/frame.rs

```diff
@@ -0,0 +1,272 @@
+use std::fmt;
+use std::mem::size_of;
+
+use bytes::{Buf, BufMut, Bytes};
+
+use quinn_proto::coding::{BufExt, BufMutExt, Codec, UnexpectedEnd};
+use quinn_proto::varint;
+
+#[derive(Debug, PartialEq)]
+pub enum Error {
+    Malformed,
+    UnsupportedFrame,
+    InvalidFrameValue,
+}
+
+#[derive(Debug, PartialEq)]
+pub enum HttpFrame {
+    Data(DataFrame),
+    Settings(SettingsFrame),
+}
+
+impl HttpFrame {
+    pub fn encode<T: BufMut>(&self, buf: &mut T) {
+        match self {
+            HttpFrame::Data(f) => f.encode(buf),
+            HttpFrame::Settings(f) => f.encode(buf),
+        }
+    }
+
+    pub fn decode<T: Buf>(buf: &mut T) -> Result<Self, Error> {
+        let len = varint::read(buf).ok_or(Error::Malformed)?;
+        let ty = buf.get::<Type>()?;
+        let mut payload = buf.take(len as usize);
+        match ty {
+            Type::DATA => Ok(HttpFrame::Data(DataFrame {
+                payload: payload.collect(),
+            })),
+            Type::SETTINGS => Ok(HttpFrame::Settings(SettingsFrame::decode(&mut payload)?)),
+            _ => Err(Error::UnsupportedFrame),
+        }
+    }
+}
+
+frame_types! {
+    DATA = 0x0,
+    HEADERS = 0x1,
+    PRIORITY = 0x2,
+    CANCEL_PUSH = 0x3,
+    SETTINGS = 0x4,
+    PUSH_PROMISE = 0x5,
+    GOAWAY = 0x7,
+    MAX_PUSH_ID = 0xD,
+}
+
+#[derive(Copy, Clone, Eq, PartialEq)]
+struct Type(u8);
+
+impl Codec for Type {
+    fn decode<B: Buf>(buf: &mut B) -> Result<Self, UnexpectedEnd> {
+        Ok(Type(buf.get_u8()))
+    }
+    fn encode<B: BufMut>(&self, buf: &mut B) {
+        buf.write(self.0);
+    }
+}
+
+trait FrameHeader {
+    fn len(&self) -> usize;
+    fn ftype(&self) -> Type;
+    fn encode_header<T: BufMut>(&self, buf: &mut T) {
+        println!("len is {}", self.len());
+        varint::write(self.len() as u64, buf).unwrap();
+        buf.put_u8(self.ftype().0);
+    }
+}
+
+#[derive(Debug, PartialEq)]
+pub struct DataFrame {
+    payload: Bytes,
```

Comment:
I'm not sure about using `Bytes` here for the interface with the user. Won't it create data copies ? I tried to use `<T> where T: AsRef<[u8]>` without success as I've seen in `quinn-proto::frame::Stream` encoding impl. I believe it's a function-call ergonomics matter as well ?

Reply:
I think using `Bytes` here is fine. The fancy polymorphic `Stream` type in `-proto` is an implementation detail, and is not actually conferring any benefit at all right now. Buffering and flow control concerns, as well as the difficulty and expense of storing generic types in non-generic data structures, generally dictate that copying occurs anyway.